tests/KubernetesClient.Tests/ThreadSafeStoreTests.cs

```diff
@@ -0,0 +1,140 @@
+            Assert.Equal(val.Kind, kind);
+
+            // Add again which should raise and exception
+            var raisedExcection = false;
+            try
+            {
+                store.Add(key, kObject);
+            }
+            catch (Exception)
+            {
+                raisedExcection = true;
+            }
+            Assert.True(raisedExcection);
+            val = store.Get(key);
+            Assert.Equal(val.ApiVersion, apiVersion);
+            Assert.Equal(val.Kind, kind);
+
+            // Update existing key
+            var newApiVersion = "v4update";
+            var newKind = "kind4update";
+            var newKObject = new KubernetesObject();
+            newKObject.ApiVersion = newApiVersion;
+            newKObject.Kind = newKind;
+
+            var oldVal = store.Update(key, newKObject);
+            // Check if the object matches the new key.
+            val = store.Get(key);
+            Assert.Equal(val.ApiVersion, newApiVersion);
+            Assert.Equal(val.Kind, newKind);
+            // Check the old value
+            Assert.Equal(oldVal.ApiVersion, apiVersion);
+            Assert.Equal(oldVal.Kind, kind);
+
+            key = "non existant key";
+            store.Update(key, kObject);
+            val = store.Get(key);
+            // Check if the updated value (by add) is correct.
+            Assert.Equal(val.ApiVersion, apiVersion);
+            Assert.Equal(val.Kind, kind);
+
+            // Delete
+            store.Delete(key);
+            val = store.Get(key);
+            Assert.Null(val);
+
+            raisedExcection = false;
+            // Delete again 
+            try
+            {
+                store.Delete(key);
+            }
+            catch (Exception)
+            {
+                raisedExcection = true;
+            }
+
+            val = store.Get(key);
+            Assert.Null(val);
+        }
+
+        [Fact]
+        public void MultiThread()
+        {
+            var store = new ThreadSafeStore();
+            Action<object> action = (object o) =>
+            {
+                var i = (int)o;
+                var key = string.Format("{0}", i);
+
+                var k = new KubernetesObject();
+                k.ApiVersion = "apiV" + key;
+                k.Kind = "kind" + key;
+                store.Add(key, k);
+            };
+
+            int max = 10;
+            var tasks = new Task[max];
+            var dict = new Dictionary<string, bool>();
+
+            for (int i = 0; i < max; i++)
+            {
+                dict.Add(i.ToString(), false);
+                tasks[i] = Task.Factory.StartNew(action, i);
+            }
+            Task.WaitAll(tasks);
+
+
+            int count = 0;
+            foreach (string key in store.ListKeys())
+            {
+                // Ensure we have not seen the key before.
+                Assert.False(dict[key] == true);
+                dict[key] = true;
```

Comment:
missing assert every single key is true in case of store.listkeys did not return everything.

just a notice, i did see `assert(count, max)` not quite sure if checking count only is good enough